### PR TITLE
K8SPXC-566 Add PITR e2e tests

### PR DESCRIPTION
**Pull request by nonemax**
[K8SPXC-566](https://jira.percona.com/browse/K8SPXC-566)

### CODE REVIEW

File: cmd/pitr/recoverer/recoverer_test.go [hunk 1]
```diff
@@ -2,70 +2,42 @@ package recoverer
...
-
-func (r *reader) Read(p []byte) (int, error) {
-	return r.r.Read(p)
-}
-
-func (t *testStorage) GetObject(name string) (io.Reader, error) {
-	if strings.Contains(name, "gtid-set") {
-		buf := bytes.NewBufferString("some-set:1-15")
-		return buf, nil
+func TestGetBucketAndPrefix(t *testing.T) {
```
Comment (dAdAbird) [inline cmd/pitr/recoverer/recoverer_test.go:8]: use table tests:
```go
func TestGetBucketAndPrefix(t *testing.T) {
	cases := []string{
		"operator-testing/test",
		"s3://operator-testing/test",
		"https://somedomain/operator-testing/test",
	}

	for _, c := range cases {
		t.Run(c, func(t *testing.T) {
			bucket, prefix, err = getBucketAndPrefix(c)
			if err != nil {
				t.Error("get from 'operator-testing/test'", err.Error())
			}
			if bucket != "operator-testing" && prefix != "test" {
				t.Error("wrong parsing of 'operator-testing/test'")
			}
		}
	})
}
```

and add more cases like
- "operator-testing/test/"
- "operator-testing/test/pitr"
- "https://somedomain/operator-testing"
- "operator-testing"
etc

File: cmd/pitr/main.go [hunk 2]
```diff
@@ -75,149 +77,31 @@ func runRecoverer() {
...
-		return collector.Config{}, errors.Wrap(err, "get BUFFER_SIZE env")
-	}
-	bufferSize, err := strconv.ParseInt(bufferSizeStr, 10, 64)
-	if err != nil {
-		return collector.Config{}, errors.Wrap(err, "get buffer size")
+	cfg := collector.Config{}
+	if err := env.Parse(&cfg); err != nil {
+		return cfg, err
 	}
+	return cfg, nil
```
Comment (dAdAbird) [inline cmd/pitr/main.go:85]:
```suggestion
	cfg := collector.Config{}
	err := env.Parse(&cfg)
	return cfg, err
```

File: cmd/pitr/main.go [hunk 3]
```diff
@@ -75,149 +77,31 @@ func runRecoverer() {
...
-	backupS3AccessKey, err := getEnv("SECRET_ACCESS_KEY", "", true)
-	if err != nil {
-		return recoverer.Config{}, errors.Wrap(err, "get SECRET_ACCESS_KEY env")
-	}
-	backupS3BackupDest, err := getEnv("S3_BUCKET_URL", "", true)
-	if err != nil {
-		return recoverer.Config{}, errors.Wrap(err, "get S3_BUCKET_URL env")
+	cfgBackupS3 := recoverer.BackupS3{}
+	if err := env.Parse(&cfgBackupS3); err != nil {
+		return cfg, err
```
Comment (dAdAbird) [inline cmd/pitr/main.go:95]: no need in tmp vars, do just:
```go
func getRecovererConfig() (recoverer.Config, error) {
	cfg := recoverer.Config{}
	if err := env.Parse(&cfg); err != nil {
		return cfg, err
	}

	if err := env.Parse(&cfgBackupS3.BackupS3); err != nil {
		return cfg, err
	}

	if err := env.Parse(&cfgBinlogS3.BinlogS3); err != nil {
		return cfg, err
	}

	return cfg, nil
}
```

**Review by dAdAbird — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| upgrade-haproxy | failed |
| haproxy | failed |
| recreate | failed |
| scaling | passed |
| storage | passed |
| big-data | failed |
| scaling-proxysql | failed |
| self-healing | passed |
| upgrade-proxysql | failed |
| restore-to-encrypted-cluster | failed |
| init-deploy | failed |
| security-context | failed |
| smart-update | failed |
| self-healing-advanced | passed |
| demand-backup | failed |
| operator-self-healing | passed |
| limits | passed |
| upgrade-consistency | failed |
| monitoring-2-0 | passed |
| scheduled-backup | failed |
| affinity | passed |
| one-pod | failed |
| demand-backup-encrypted-with-tls | failed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | failed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | failed |
| scaling | passed |
| recreate | failed |
| big-data | failed |
| upgrade-haproxy | failed |
| storage | passed |
| scaling-proxysql | failed |
| self-healing | passed |
| restore-to-encrypted-cluster | failed |
| init-deploy | failed |
| security-context | failed |
| demand-backup | failed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| scheduled-backup | failed |
| one-pod | failed |
| auto-tuning | passed |
| self-healing-advanced | passed |
| demand-backup-encrypted-with-tls | failed |
| proxysql-sidecar-res-limits | passed |
| users | failed |
| operator-self-healing | passed |
| pitr | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| upgrade-haproxy | failed |
| recreate | failed |
| haproxy | failed |
| storage | passed |
| scaling | passed |
| self-healing | passed |
| scaling-proxysql | failed |
| big-data | failed |
| self-healing-advanced | passed |
| operator-self-healing | passed |
| upgrade-proxysql | failed |
| restore-to-encrypted-cluster | failed |
| security-context | failed |
| init-deploy | failed |
| smart-update | failed |
| demand-backup | failed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | failed |
| upgrade-consistency | failed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | failed |
| scheduled-backup | failed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| demand-backup-encrypted-with-tls | failed |
| pitr | passed |

File: e2e-tests/pitr/conf/restore-on-pitr-minio.yaml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+#      destination: s3://operator-testing/demand-backup-2020-12-03-09:05:22-full
+#      s3:
+#        bucket: operator-testing
+#        credentialsSecret: gcp-cs-secret
+#        endpointUrl: https://storage.googleapis.com
+#        region: us-east-1
+  pitr:
+    type: latest
+    backupSource:
+      storageName: "minio-binlogs"
```
Comment (delgod) [inline e2e-tests/pitr/conf/restore-on-pitr-minio.yaml:18]: can we use `backupSource` approach for both backup and PITR here?
Reason: to be sure that we can recover from backup to another k8s cluster

File: e2e-tests/pitr/conf/pitr-job.yaml [hunk 5]
```diff
@@ -0,0 +1,89 @@
+apiVersion: batch/v1
```
Comment (delgod) [inline e2e-tests/pitr/conf/pitr-job.yaml:1]: why this file needed?

File: e2e-tests/pitr/run [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+    desc 'recover backup' $restore
+    kubectl_bin apply -f "$test_dir/conf/${restore}.yaml"
+    wait_backup_restore ${backup}
+    kubectl patch pxc pitr --type=merge -p "{\"spec\":{\"backup\":{\"pitr\":{\"enabled\":false}}}}"
+    kubectl_bin logs job/restore-job-${backup}-${cluster}
+    kubectl_bin delete -f "$test_dir/conf/${restore}.yaml"
+    wait_for_running "$cluster-proxysql" 1
+    wait_for_running "$cluster-pxc" 3
+    kubectl patch pxc pitr --type=merge -p "{\"spec\":{\"backup\":{\"pitr\":{\"enabled\":false}}}}"
```
Comment (delgod) [inline e2e-tests/pitr/run:24]: why do we need to call it 3 times?

**Review by delgod — CHANGES_REQUESTED** [review]

File: e2e-tests/pitr/run (hunk 6 above)
Comment (nonemax) [reply e2e-tests/pitr/run]: fixed https://github.com/percona/percona-xtradb-cluster-operator/pull/697/commits/867e25afaa236becf1253b02c8bf67e3c397219e

File: e2e-tests/pitr/conf/pitr-job.yaml (hunk 5 above)
Comment (nonemax) [reply e2e-tests/pitr/conf/pitr-job.yaml]: removed https://github.com/percona/percona-xtradb-cluster-operator/pull/697/commits/867e25afaa236becf1253b02c8bf67e3c397219e

File: e2e-tests/pitr/conf/restore-on-pitr-minio.yaml (hunk 4 above)
Comment (nonemax) [reply e2e-tests/pitr/conf/restore-on-pitr-minio.yaml]: done https://github.com/percona/percona-xtradb-cluster-operator/pull/697/commits/867e25afaa236becf1253b02c8bf67e3c397219e

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

**Review by dAdAbird — DISMISSED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| upgrade-haproxy | passed |
| scaling | passed |
| haproxy | passed |
| recreate | passed |
| big-data | passed |
| self-healing | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| limits | passed |
| self-healing-advanced | passed |
| monitoring-2-0 | passed |
| demand-backup | failed |
| smart-update | failed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| demand-backup-encrypted-with-tls | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| recreate | passed |
| big-data | passed |
| upgrade-haproxy | passed |
| scaling | passed |
| haproxy | passed |
| init-deploy | passed |
| self-healing | failed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| limits | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| smart-update | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| self-healing-advanced | passed |
| proxysql-sidecar-res-limits | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| users | passed |
| upgrade-consistency | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| pitr | passed |

**Comment by JNKPercona** [comment]
PXC operator docker - perconalab/percona-xtradb-cluster-operator:PR-697-1025d228

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| recreate | passed |
| big-data | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-encrypted-with-tls | passed |
| pitr | passed |
| upgrade-haproxy | passed |
| storage | passed |
| self-healing | passed |
| haproxy | passed |
| init-deploy | passed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |
| self-healing-advanced | passed |
| operator-self-healing | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| upgrade-haproxy | passed |
| recreate | passed |
| scaling | passed |
| storage | passed |
| big-data | passed |
| haproxy | passed |
| self-healing | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| security-context | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| limits | passed |
| monitoring-2-0 | passed |
| smart-update | passed |
| demand-backup | passed |
| affinity | passed |
| one-pod | passed |
| self-healing-advanced | passed |
| auto-tuning | passed |
| operator-self-healing | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| scheduled-backup | failed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

**Comment by JNKPercona** [comment]
PXC operator docker - perconalab/percona-xtradb-cluster-operator:PR-697-1f9cfdd1

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| storage | passed |
| self-healing | passed |
| self-healing-advanced | passed |
| operator-self-healing | passed |
| recreate | passed |
| scaling | passed |
| upgrade-haproxy | passed |
| restore-to-encrypted-cluster | passed |
| scaling-proxysql | passed |
| security-context | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| haproxy | passed |
| init-deploy | passed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |
| demand-backup-encrypted-with-tls | passed |
| pitr | passed |

File: e2e-tests/pitr/conf/restore-on-pitr-minio.yaml [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+      destination: <destination>
+      s3:
+        bucket: operator-testing
+        credentialsSecret: minio-secret
+        endpointUrl: http://minio-service:9000
+        region: us-east-1
+  pitr:
+    type: latest
+    backupSource:
+      storageName: "minio-binlogs"
```
Comment (delgod) [inline e2e-tests/pitr/conf/restore-on-pitr-minio.yaml:19]: please don't specify storage and specify only s3 details
```suggestion
  backupSource:
    destination: <destination>
    s3:
      bucket: operator-testing
      credentialsSecret: minio-secret
      endpointUrl: http://minio-service:9000
      region: us-east-1
  pitr:
    type: latest
    backupSource:
      s3:
        bucket: operator-testing/binlogs
        credentialsSecret: minio-secret
        endpointUrl: http://minio-service:9000
        region: us-east-1
```

File: Jenkinsfile [hunk 8]
```diff
@@ -123,6 +123,8 @@ pipeline {
         VERSION = "${env.GIT_BRANCH}-${env.GIT_SHORT_COMMIT}"
         CLUSTER_NAME = sh(script: "echo jenkins-pxc-${GIT_SHORT_COMMIT} | tr '[:upper:]' '[:lower:]'", , returnStdout: true).trim()
         AUTHOR_NAME  = sh(script: "echo ${CHANGE_AUTHOR_EMAIL} | awk -F'@' '{print \$1}'", , returnStdout: true).trim()
+        IMAGE_PXC="perconalab/percona-xtradb-cluster:8.0.20-11-2_custom"
+        IMAGE_BACKUP="perconalab/percona-xtradb-cluster-operator:k8spxc-566-backup"
```
[the remark this answers is not in the document]
Comment (delgod) [reply Jenkinsfile:128]: all image changes are merged, we can remove these lines

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| upgrade-haproxy | passed |
| haproxy | passed |
| big-data | passed |
| scaling | passed |
| recreate | passed |
| self-healing | failed |
| init-deploy | failed |
| scaling-proxysql | passed |
| security-context | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| limits | passed |
| demand-backup | passed |
| smart-update | passed |
| monitoring-2-0 | passed |
| self-healing-advanced | passed |
| affinity | passed |
| one-pod | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| scheduled-backup | failed |
| proxysql-sidecar-res-limits | passed |
| users | failed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| pitr | failed |

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| scaling | passed |
| big-data | passed |
| upgrade-haproxy | passed |
| haproxy | passed |
| recreate | passed |
| self-healing | failed |
| init-deploy | failed |
| scaling-proxysql | passed |
| security-context | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| limits | passed |
| demand-backup | passed |
| smart-update | passed |
| monitoring-2-0 | passed |
| self-healing-advanced | passed |
| affinity | passed |
| one-pod | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| scheduled-backup | failed |
| users | failed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| demand-backup-encrypted-with-tls | failed |
| pitr | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| storage | passed |
| haproxy | passed |
| big-data | passed |
| recreate | passed |
| upgrade-haproxy | passed |
| init-deploy | failed |
| scaling-proxysql | passed |
| self-healing | failed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| limits | passed |
| demand-backup | passed |
| smart-update | passed |
| monitoring-2-0 | passed |
| self-healing-advanced | passed |
| affinity | passed |
| one-pod | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| scheduled-backup | failed |
| proxysql-sidecar-res-limits | passed |
| users | failed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| demand-backup-encrypted-with-tls | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| big-data | passed |
| haproxy | passed |
| recreate | passed |
| scaling | passed |
| upgrade-haproxy | passed |
| init-deploy | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| limits | passed |
| smart-update | passed |
| demand-backup | passed |
| self-healing-advanced | passed |
| monitoring-2-0 | passed |
| operator-self-healing | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| scheduled-backup | passed |
| proxysql-sidecar-res-limits | passed |
| upgrade-consistency | passed |
| users | passed |
| demand-backup-encrypted-with-tls | failed |
| tls-issue-self | failed |
| pitr | failed |
| tls-issue-cert-manager | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| recreate | passed |
| big-data | passed |
| haproxy | passed |
| storage | passed |
| upgrade-haproxy | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| self-healing | passed |
| security-context | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| limits | passed |
| self-healing-advanced | passed |
| operator-self-healing | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| demand-backup | passed |
| one-pod | passed |
| scheduled-backup | passed |
| auto-tuning | passed |
| demand-backup-encrypted-with-tls | failed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | failed |
| pitr | failed |

**Comment by JNKPercona** [comment]
PXC operator docker - perconalab/percona-xtradb-cluster-operator:PR-697-6555e062

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| scaling-proxysql | passed |
| security-context | passed |
| upgrade-haproxy | passed |
| big-data | passed |
| storage | passed |
| self-healing | passed |
| recreate | passed |
| self-healing-advanced | passed |
| restore-to-encrypted-cluster | passed |
| operator-self-healing | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-encrypted-with-tls | passed |
| haproxy | passed |
| init-deploy | passed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | passed |
| upgrade-proxysql | passed |
| tls-issue-cert-manager | passed |
| smart-update | passed |
| upgrade-consistency | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| pitr | passed |

**Pull request merged**